### PR TITLE
Packages: Fix all missing or obsolete dependencies in packages

### PR DESCRIPTION
**Pull request by gziolo**
## Description
I discovered that we have some differences in the list of dependencies for WordPress packages. In particular, I have some issues integrating all recently published npm packages with WordPress core. Related ticket:
https://core.trac.wordpress.org/ticket/47843

## How has this been tested?
In general, on Gutenberg side, it all should be transparent as we have build tools which properly generate dependencies consumed by WordPress. In core, we have to do it manually so that might be why it doesn't quite work as necessary.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -106,6 +109,16 @@ module.exports = {
 		} ],
 	},
 	overrides: [
+		{
+			files: [ 'packages/**/*.js' ],
+			rules: {
+				'import/no-extraneous-dependencies': 'error',
+			},
+			excludedFiles: [
+				'**/*.@(android|ios|native).js',
```
Comment (gziolo) [inline .eslintrc.js]: @hypest and @koke - I disabled native files for now to fix all the issues which web version has, but we definitely should revisit it really soon.

**Comment by ellatrix** [comment]
I wonder if there's any way we can check the JS files for imports and error if the dependency is not found in the `package.json` file, and also error if there are dependencies that are listed and not used.

**Comment by gziolo** [comment]
> I wonder if there's any way we can check the JS files for imports and error if the dependency is not found in the `package.json` file, and also error if there are dependencies that are listed and not used.

So this new ESLint rule (`import/no-extraneous-dependencies`) I added in this PR solves the first case :) 
I couldn't find anything which covers 2nd scenario.

File: packages/block-editor/package.json [hunk 2]
```diff
@@ -28,6 +28,7 @@
 		"@wordpress/blocks": "file:../blocks",
 		"@wordpress/components": "file:../components",
 		"@wordpress/compose": "file:../compose",
+		"@wordpress/core-data": "file:../core-data",
```
Comment (youknowriad) [inline packages/block-editor/package.json]: Why do we need this? This package shouldn't depend on core data?

Comment (gziolo) [reply packages/block-editor/package.json]: It comes from:

https://github.com/WordPress/gutenberg/blob/1c7794121a74fb8458e660f5c521c4915a69fffb/packages/block-editor/src/index.js#L5

Comment (gziolo) [reply packages/block-editor/package.json]: I don't see it used so it's probably copied over from `@wordpress/editor`.

Comment (gziolo) [reply packages/block-editor/package.json]: Fix ready: 8a017ee.

**Comment by ellatrix** [comment]
> I couldn't find anything which covers 2nd scenario.

How did you know which dependencies to remove then?

**Review by ellatrix — APPROVED** [review]
Awesome!

**Comment by gziolo** [comment]
> > I couldn't find anything which covers 2nd scenario.
> 
> How did you know which dependencies to remove then?

It was a manual process. I went through `build` folder where we have JSON files generated with dependencies for each module which are used in the code. Based on that, I removed those WP packages which weren't listed. It would be great to have this automated with a linter rule.

File: .eslintrc.js (hunk 1 above)
Comment (gziolo) [reply .eslintrc.js]: there is a follow-up: #16978

**Pull request merged**